### PR TITLE
Add support of backward_prefetch and forward_prefetch

### PR DESCRIPTION
**Pull request by raghavanone**
#21156 

Adds support for backward_prefetch and forward_prefetch in trainer.

@sgugger @pacman100

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -417,6 +417,15 @@ def __init__(
             elif FSDPOption.NO_SHARD in args.fsdp:
                 self.fsdp = ShardingStrategy.NO_SHARD
 
+            self.backward_prefetch = BackwardPrefetch.BACKWARD_PRE
+
+            if "backward_prefetch" in args.fsdp_backward_prefetch:
```
Comment (pacman100) [inline src/transformers/trainer.py:422]:
```suggestion
            if "backward_prefetch" not in args.fsdp_backward_prefetch:
```

File: src/transformers/training_args.py [hunk 2]
```diff
@@ -408,6 +408,19 @@ class TrainingArguments:
             - `"auto_wrap"`: Automatically recursively wrap layers with FSDP using `default_auto_wrap_policy`.
         fsdp_min_num_params (`int`, *optional*, defaults to `0`):
             FSDP's minimum number of parameters for Default Auto Wrapping. (useful only when `fsdp` field is passed).
+        fsdp_backward_prefetch (`str`, *optional*)
+            FSDP's backward prefetch mode. Controls when to prefetch next set of parameters
```
Comment (pacman100) [inline src/transformers/training_args.py:412]:
```suggestion
            FSDP's backward prefetch mode. Controls when to prefetch next set of parameters (useful only when `fsdp` field is passed).
```

File: src/transformers/training_args.py [hunk 3]
```diff
@@ -408,6 +408,19 @@ class TrainingArguments:
...
+
+            A list of options along the following:
+
+            - `"backward_pre"` : Prefetches the next set of parameters before the current set of parameter's gradient
+                computation.
+            - `"backward_pos"` : This prefetches the next set of parameters after the current set of parameter’s
+                gradient computation.
+        fsdp_forward_prefetch (`bool`, *optional*, defaults to `False`)
+            FSDP's backward prefetch mode.
```
Comment (pacman100) [inline src/transformers/training_args.py:421]:
```suggestion
            FSDP's forward prefetch mode (useful only when `fsdp` field is passed).
```

File: src/transformers/training_args.py [hunk 4]
```diff
@@ -866,6 +879,24 @@ class TrainingArguments:
             ),
         },
     )
+    fsdp_backward_prefetch: str = field(
+        default="backward_pre",
+        metadata={
+            "help": (
+                "Which  backward_prefetch mode to use of 'backward_pre' or 'backward post' which using PyTorch Fully "
```
Comment (pacman100) [inline src/transformers/training_args.py:886]:
```suggestion
                "Which  `backward_prefetch` mode to use out of 'backward_pre' or 'backward post' when using PyTorch Fully "
```

**Review by pacman100 — COMMENTED** [review]
Hello @raghavanone,  Thank you for the additions 🤗! Left a few comments/suggestions.

**Review by sgugger — COMMENTED** [review]
I don't think we can keep adding parameters like this that are only useful to the FSDP integration. Ultimately we will have the Accelerate config that will allow us to customize them.

In the meantime, I think the right argument to add is a path to a fsdp_config and deprecate `fsdp_min_num_params`, then put all these arguments in this config.

**Comment by raghavanone** [comment]
@sgugger Done, But not sure why this test is failing. Any pointers on how to make this build green would help.

File: src/transformers/trainer.py [hunk 5]
```diff
@@ -417,13 +420,20 @@ def __init__(
...
                 self.fsdp = ShardingStrategy.NO_SHARD
 
-            self.backward_prefetch = BackwardPrefetch.BACKWARD_PRE
+            if isinstance(self.args.fsdp_config, dict):
+                self.fsdp_config = self.args.fsdp_config
+            elif isinstance(self.args.fsdp_config, str):
+                with io.open(self.args.fsdp_config, "r", encoding="utf-8") as f:
+                    self.fsdp_config = json.load(f)
+            else:
+                self.fsdp_config = {}
```
Comment (sgugger) [inline src/transformers/trainer.py:428]: This is super nice, but should be done in the post init of the TrainingArguments :-)

File: src/transformers/training_args.py [hunk 6]
```diff
@@ -407,21 +407,28 @@ class TrainingArguments:
               `"shard_grad_op"`).
             - `"auto_wrap"`: Automatically recursively wrap layers with FSDP using `default_auto_wrap_policy`.
         fsdp_min_num_params (`int`, *optional*, defaults to `0`):
```
Comment (sgugger) [inline src/transformers/training_args.py:409]: You can remove the documentation of the deprecated argument, but make sure that when the user sets it, we issue a deprecation warning and update the fsdp_config (in the post init).

File: src/transformers/trainer.py [hunk 7]
```diff
@@ -1406,10 +1416,14 @@ def _wrap_model(self, model, training=True, dataloader=None):
                 cpu_offload = CPUOffload(offload_params=False)
 
             auto_wrap_policy = None
+
+            self.fsdp_min_num_params = self.args.fsdp_min_num_params
```
Comment (sgugger) [inline src/transformers/trainer.py:1419]: SHould be dealt with in the post init.

**Review by sgugger — COMMENTED** [review]
Thanks a lot for iterating on this! I left additional comments but this is way better! Could you also show an example in the FSDP doc page?

File: src/transformers/training_args.py [hunk 8]
```diff
@@ -866,12 +890,39 @@ class TrainingArguments:
...
+    )
+    fsdp_forward_prefetch: Optional[bool] = field(
+        default=False,
+        metadata={
+            "help": (
+                "If True, then FSDP explicitly prefetches the next upcoming all-gather"
+                "while executing in the forward pass"
+            ),
+        },
+    )
```
Comment (pacman100) [inline src/transformers/training_args.py:919]:
```suggestion
```

File: src/transformers/training_args.py [hunk 9]
```diff
@@ -866,12 +890,39 @@ class TrainingArguments:
             ),
         },
     )
+    fsdp_config: Optional[str] = field(
+        default=None,
+        metadata={
+            "help": (
+                "Config to be used with fsdp (Pytorch Distributed Parallel Training). The  value is either the of "
+                "deepspeed json config file (e.g., `ds_config.json`) or an already loaded  json file as `dict`."
```
Comment (pacman100) [inline src/transformers/training_args.py:898]:
```suggestion
                "fsdp json config file (e.g., `fsdp_config.json`) or an already loaded  json file as `dict`."
```

File: src/transformers/training_args.py [hunk 10]
```diff
@@ -866,12 +890,39 @@ class TrainingArguments:
             ),
         },
     )
+    fsdp_config: Optional[str] = field(
+        default=None,
+        metadata={
+            "help": (
+                "Config to be used with fsdp (Pytorch Distributed Parallel Training). The  value is either the of "
```
Comment (pacman100) [inline src/transformers/training_args.py:897]:
```suggestion
                "Config to be used with FSDP (Pytorch Fully Sharded Data Parallel Training). The  value is either a"
```

**Review by pacman100 — COMMENTED** [review]
Hello, Thank you for iterating to support FSDP config file.
Left few suggestions.

**Review by pacman100 — COMMENTED** [review]
Hello, Thank you for iterating to support FSDP config file.
Left few suggestions.

**Comment by raghavanone** [comment]
@sgugger @pacman100  Need pointer on why this test is failing.

File: src/transformers/training_args.py [hunk 11]
```diff
@@ -1277,13 +1327,28 @@ def __post_init__(self):
...
+        if self.fsdp_config is None:
+            self.fsdp_config = {}
+
+        if isinstance(self.fsdp_config, str):
+            with io.open(self.fsdp_config, "r", encoding="utf-8") as f:
+                self.fsdp_config = json.load(f)
+
+        self.fsdp_config["fsdp_min_num_params"] = max(
+            getattr(self.fsdp_config, "fsdp_min_num_params", 0), self.fsdp_min_num_params
+        )
```
Comment (sgugger) [inline src/transformers/training_args.py:1339]: Missing the deprecation warning for `self.fsdp_min_num_params > 0` here.

**Review by sgugger — COMMENTED** [review]
The test is a flaky one, don't worry about it. Thanks for iterating, I just have one last comment on the deprecation warning for `fsdp_min_num_params` and we can merge this!

File: src/transformers/training_args.py (hunk 11 above)
Comment (raghavanone) [reply src/transformers/training_args.py:1339]: Done.

Comment (sgugger) [reply src/transformers/training_args.py:1339]: I think you forgot to push the corresponding update?

**Comment by raghavanone** [comment]
> The test is a flaky one, don't worry about it. Thanks for iterating, I just have one last comment on the deprecation warning for `fsdp_min_num_params` and we can merge this!

Done

File: src/transformers/training_args.py (hunk 8 above)
Comment (pacman100) [reply src/transformers/training_args.py:927]: You haven't removed these fields. These aren't required now as they are handled in the `fsdp_config`

File: src/transformers/training_args.py (hunk 9 above)
Comment (pacman100) [reply src/transformers/training_args.py:906]: This is yet to be done

File: src/transformers/training_args.py [hunk 12]
```diff
@@ -870,11 +893,38 @@ class TrainingArguments:
...
+                "This parameter is deprecatetd. FSDP's minimum number of parameters for Default Auto Wrapping. (useful"
+                " only when `fsdp` field is passed)."
             )
         },
     )
+    fsdp_config: Optional[str] = field(
+        default=None,
+        metadata={
+            "help": (
+                "Config to be used with FSDP (Pytorch Fully Sharded  Data Parallel). The  value is either the of "
```
Comment (pacman100) [inline src/transformers/training_args.py:905]:
```suggestion
                "Config to be used with FSDP (Pytorch Fully Sharded  Data Parallel). The  value is either the of
```suggestion
                "Config to be used with FSDP (Pytorch Fully Sharded  Data Parallel). The  value is either a "
```
 "
```

Comment (raghavanone) [reply src/transformers/training_args.py]: Done. Can we merge it now ?

**Comment by pacman100** [comment]
?Hello @raghavanone , could you please resolve the comments above that I have unresolved as they are yet to be addressed ?

**Comment by raghavanone** [comment]
> ?Hello @raghavanone , could you please resolve the comments above that I have unresolved as they are yet to be addressed ?

Done

File: src/transformers/training_args.py (hunk 8 above)
Comment (sgugger) [reply src/transformers/training_args.py:927]: This still needs to be addressed.

File: src/transformers/training_args.py (hunk 11 above)
Comment (sgugger) [reply src/transformers/training_args.py:1339]: This has not been addressed.

**Review by sgugger — COMMENTED** [review]
@raghavanone Please address all comments before pinging us for a re-review, and only resolve comments in the GitHub interface once they have been addressed.

**Review by sgugger — APPROVED** [review]
Thanks a lot for your contribution!

**Comment by pacman100** [comment]
Thank you @raghavanone for iterating and addressing the comment and for the overall contribution! 🚀

**Pull request merged**